# Instructions

1. Research and find a suitable dataset for your term project. The dataset should be rich enough to allow for interesting analysis and visualizations.
2. Once you have located your dataset, download it and import it into a Python environment in Jupyter Notebook.
3. The first section of your notebook should include why you chose this data and data provenance.
4. Use the pandas library to clean and prepare your data. This may include dealing with missing values, converting data types, or restructuring the data as necessary. By the end your data should be [‘tidy’ data.](https://vita.had.co.nz/papers/tidy-data.pdf)
5. After your data is cleaned, complete an exploratory analysis of your data. Depending on your data, this may include things like:
    1. How many observations do you have?
    2. How many unique categories are there for categorical variables?
    3. How much missing data is there per observation or per variable?
    4. What are the distributions of continuous variables? Are there outliers?
    
    There may be more or different areas depending on your data. Using code chunks and RMarkdown chunks liberally. [This link](https://www.analyticsvidhya.com/blog/2021/04/20-must-known-pandas-function-for-exploratory-data-analysis-eda/) shows some starter functions to use. 
    
6. At the end of your data exploration, include a data dictionary. 
7. Finally, brainstorm a list of UI components that you could include in your dashboard. Consider what would be most useful or interesting for users.
8. Also brainstorm a list 3-6 possible data visualizations (e.g., a scatter plot of sepal length vs sepal width - this could be filtered on species using a multi-select dropdown)

# **Data Provenance**



***Why I chose this dataset***

I chose this dataset because I have always been interested in facility emissions and the types of emissions (i.e. direct emissions, gathering & boosting, etc.) that various facilities emit. Additionally, my future career will be focusing on EPA Energy Star and other facility emission reporting, so I thought that using this dataset in my project will help me gather useful information that will be helpful in the long term. 

***Data background information***

***Where the data came from:***
This dataset was pulled from the EPA's Greenhouse Gas Reporting Program (or GHGRP) for the reporting year of 2022. This program collects data from approximately 8,000 facilities that are required to report their emissions annually; these facilities are typically large GHG emission sources, fuel and industrial gas suppliers, and carbon dioxide injection sites. The data is compiled every April and made available to the public the following October after a thorough multi-step verification process, where the EPA ensures that the reported data is accurate, complete, and consistent. 

***Original purpose of data:***
The data was originally used to keep these 8,000 high-emitting facilities in check through annual emissions reporting. Additionally, the EPA allowed this data to be public information for other businesses to track and compare different facilities' greenhouse gas emission as to find ways to cut down their own emissions, minimize wasted resources, and (subsequently) save money. Moreover, the data was also created for local governments to find high-emitting facilities in their area to develop effective climate policies. 

***Data overview:***
The data contains report for direct emitting facilities (Scope 1) and upstream suppliers in 2022. Scope 1 emissions are reported at the individual facility level, though parent company information is also collected. Regarding upstream suppliers, they report the amount of carbon dioxide equivalent (CO2e) that would be hypothetically released if their products were released, combusted, or oxidized; these emissions are reported at the corporate level and fall under Scope 3 emissions. The data collected by the GHGRP cover approximately 85-90% of US greenhouse gas emissions. However, the GHGRP does not include emissions from agriculture, emission sources that have annual emissions of less than 25,000 metric tons of CO2e, sinks of greenhouse gases, and Scope 2 emissions (electricty purchases or indirect emissions from energy consumption). Facilities that are required to report to the GHGRP include facilities that have GHG emissions which exceed 25,0000 metric tons of CO2e per year, contain supplies of products that would result in over 25,000 metric tons of CO2e if products were released, and/or receive 25,0000 metric tons or more of CO2 for underground injection. 

# **Data Dictionary**

# **Cleaning Data**

In [1]:
# importing dependencies 
import pandas as pd 

In [4]:
# have to install openpyxl to use pandas for reading in excel file 
pip install openpyxl

  Using cached openpyxl-3.1.2-py2.py3-none-any.whl.metadata (2.5 kB)
  Using cached et_xmlfile-1.1.0-py3-none-any.whl.metadata (1.8 kB)
Using cached openpyxl-3.1.2-py2.py3-none-any.whl (249 kB)
Using cached et_xmlfile-1.1.0-py3-none-any.whl (4.7 kB)
Note: you may need to restart the kernel to use updated packages.


***getting dataset from excel file***

In [2]:
# since there are multiple sheets in the excel file, using ExcelFile function to first read in the whole thing
ghg_data = pd.ExcelFile("/Users/sophiekim/Desktop/4th year (spring)/DS 4003/project/ghgp_data_2022.xlsx")

# then using read_excel function to read in each sheet seperately

# this one is for the directly emitting facilities 
dir_emit_data = pd.read_excel(ghg_data, 'Direct Emitters')

# this one is for the onshore gas & oil producing facilities 
onshore_gas_oil_data = pd.read_excel(ghg_data, 'Onshore Oil & Gas Prod.')

# this one is for the gathering & boosting facilities 
gat_boost_data = pd.read_excel(ghg_data, 'Gathering & Boosting')

# this one is for the transmission pipelines 
tran_pipe_data = pd.read_excel(ghg_data, 'Transmission Pipelines')

# this one is for local distribution companies' (LDC) direct emissions 
ldc_data = pd.read_excel(ghg_data, 'LDC - Direct Emissions')

# this one is for electrical equipment emissions in facilities (emitting SF6 - or sulfur hexafluoride)
sf6_elec_data = pd.read_excel(ghg_data, 'SF6 from Elec. Equip.')

# this one is for industry type 
indust_type_data = pd.read_excel(ghg_data, 'Industry Type')

# supplier facilities and CO2e injection facilities were not included because there was a lot of data missing from both these sheets 
# the data missing was significant enough that I felt that these sheets would not provide thorough enough information


***using head function to get glimpse into each df***

In [13]:
# taking a look at each df
dir_emit_data.head()

Summary data collected by the Greenhouse Gas Reporting Program for 2022  \
0  This data was reported to EPA by facilities as...                        
1  All emissions data is presented in units of me...                        
2                                        Facility Id                        
3                                            1002685                        
4                                            1004377                        

     Unnamed: 1                      Unnamed: 2 Unnamed: 3 Unnamed: 4  \
0           NaN                             NaN        NaN        NaN   
1           NaN                             NaN        NaN        NaN   
2        FRS Id                   Facility Name       City      State   
3  110071162081                        AC 857 A   Offshore         TX   
4  110043803578  121 REGIONAL DISPOSAL FACILITY    MELISSA         TX   

  Unnamed: 5                Unnamed: 6     Unnamed: 7 Unnamed: 8 Unnamed: 9  \
0        NaN                       NaN            NaN        NaN        NaN   
1        NaN                       NaN            NaN        NaN        NaN   
2   Zip Code                   Address         County   Latitude  Longitude   
3      00000                       NaN            NaN  26.295556 -94.897778   
4      75454  3820 SAM RAYBURN HIGHWAY  COLLIN COUNTY   33.29857  -96.53586   

   ...                  Unnamed: 56             Unnamed: 57      Unnamed: 58  \
0  ...                          NaN                     NaN              NaN   
1  ...                          NaN                     NaN              NaN   
2  ...  Titanium Dioxide Production  Underground Coal Mines  Zinc Production   
3  ...                          NaN                     NaN              NaN   
4  ...                          NaN                     NaN              NaN   

           Unnamed: 59                      Unnamed: 60  \
0                  NaN                              NaN   
1                  NaN                              NaN   
2  Municipal Landfills  Industrial Wastewater Treatment   
3                  NaN                              NaN   
4            325416.25                              NaN   

                                         Unnamed: 61  \
0                                                NaN   
1                                                NaN   
2  Manufacture of Electric Transmission and Distr...   
3                                                NaN   
4                                                NaN   

                  Unnamed: 62  \
0                         NaN   
1                         NaN   
2  Industrial Waste Landfills   
3                         NaN   
4                         NaN   

                                         Unnamed: 63  \
0                                                NaN   
1                                                NaN   
2  Is some CO2 collected on-site and used to manu...   
3                                                  N   
4                                                  N   

                                         Unnamed: 64  \
0                                                NaN   
1                                                NaN   
2  Is some CO2 reported as emissions from the aff...   
3                                                  N   
4                                                  N   

                                         Unnamed: 65  
0                                                NaN  
1                                                NaN  
2  Does the facility employ continuous emissions ...  
3                                                  N  
4                                                  N  

[5 rows x 66 columns]

In [14]:
onshore_gas_oil_data.head()

Summary data collected by the Greenhouse Gas Reporting Program for 2022  \
0  This data was reported to EPA by facilities as...                        
1  All emissions data is presented in units of me...                        
2                                        Facility Id                        
3                                            1000355                        
4                                            1009238                        

  Unnamed: 1                              Unnamed: 2  \
0        NaN                                     NaN   
1        NaN                                     NaN   
2     FRS Id                           Facility Name   
3        NaN  (540) Civitas Resources - Denver Basin   
4        NaN                220 Gulf Coast Basin DEC   

                        Unnamed: 3     Unnamed: 4      Unnamed: 5  \
0                              NaN            NaN             NaN   
1                              NaN            NaN             NaN   
2                            Basin  Reported City  Reported State   
3               540 - Denver Basin         Denver              CO   
4  220 - Gulf Coast Basin (LA, TX)  Oklahoma City              OK   

          Unnamed: 6               Unnamed: 7       Unnamed: 8  \
0                NaN                      NaN              NaN   
1                NaN                      NaN              NaN   
2  Reported Zip Code         Reported Address  Reported County   
3              80202  555 17th St. Suite 3700              NaN   
4              73102    333 West Sheridan Ave              NaN   

          Unnamed: 9         Unnamed: 10         Unnamed: 11  \
0                NaN                 NaN                 NaN   
1                NaN                 NaN                 NaN   
2  Reported Latitude  Reported Longitude  Primary NAICS Code   
3           39.74431          -104.98858              211120   
4          35.466697           -97.51453              211120   

                Unnamed: 12  \
0                       NaN   
1                       NaN   
2  Industry Type (subparts)   
3                    W-ONSH   
4                    W-ONSH   

                                         Unnamed: 13  \
0                                                NaN   
1                                                NaN   
2  Total reported emissions from Onshore Oil & Ga...   
3                                        1171208.898   
4                                        1173336.492   

                                         Unnamed: 14  \
0                                                NaN   
1  Total Emissions from Onshore Oil & Gas Product...   
2                      CO2 emissions (non-biogenic)    
3                                           638405.5   
4                                           913156.4   

                Unnamed: 15                     Unnamed: 16  \
0                       NaN                             NaN   
1                       NaN                             NaN   
2  Methane (CH4) emissions   Nitrous Oxide (N2O) emissions    
3                  532296.5                         506.898   
4                 259679.75                         500.342   

                                         Unnamed: 17  
0                                                NaN  
1                                                NaN  
2  Does the facility employ continuous emissions ...  
3                                                  N  
4                                                  N

In [15]:
gat_boost_data.head() 

Summary data collected by the Greenhouse Gas Reporting Program for 2022  \
0  This data was reported to EPA by facilities as...                        
1  All emissions data is presented in units of me...                        
2                                        Facility Id                        
3                                            1014558                        
4                                            1012507                        

     Unnamed: 1                                   Unnamed: 2  \
0           NaN                                          NaN   
1           NaN                                          NaN   
2        FRS Id                                Facility Name   
3           NaN  (540) Civitas Resources - GB - Denver Basin   
4  110070082329                         220 Gulf Coast Basin   

                        Unnamed: 3     Unnamed: 4      Unnamed: 5  \
0                              NaN            NaN             NaN   
1                              NaN            NaN             NaN   
2                            Basin  Reported City  Reported State   
3               540 - Denver Basin         Denver              CO   
4  220 - Gulf Coast Basin (LA, TX)        Houston              TX   

          Unnamed: 6                 Unnamed: 7       Unnamed: 8  \
0                NaN                        NaN              NaN   
1                NaN                        NaN              NaN   
2  Reported Zip Code           Reported Address  Reported County   
3              80202    555 17th St. Suite 3700              NaN   
4              77002  811 Louisiana, Suite 2100    HARRIS COUNTY   

          Unnamed: 9         Unnamed: 10         Unnamed: 11  \
0                NaN                 NaN                 NaN   
1                NaN                 NaN                 NaN   
2  Reported Latitude  Reported Longitude  Primary NAICS Code   
3          39.745822         -104.989243              211120   
4           29.75993          -95.366413              211130   

                Unnamed: 12  \
0                       NaN   
1                       NaN   
2  Industry Type (subparts)   
3                      W-GB   
4                      W-GB   

                                         Unnamed: 13  \
0                                                NaN   
1                                                NaN   
2  Total reported emissions from Gathering & Boos...   
3                                          79593.288   
4                                      228165.616365   

                                        Unnamed: 14               Unnamed: 15  \
0                                               NaN                       NaN   
1  Total Emissions from Gathering & Boosting by Gas                       NaN   
2                     CO2 emissions (non-biogenic)   Methane (CH4) emissions    
3                                           73618.7                      5943   
4                                     201806.334775               26249.79805   

                      Unnamed: 16  
0                             NaN  
1                             NaN  
2  Nitrous Oxide (N2O) emissions   
3                          31.588  
4                       109.48354

In [16]:
tran_pipe_data.head()

Summary data collected by the Greenhouse Gas Reporting Program for 2022  \
0  This data was reported to EPA by facilities as...                        
1  All emissions data is presented in units of me...                        
2                                        Facility Id                        
3                                            1013856                        
4                                            1012634                        

     Unnamed: 1                                        Unnamed: 2  \
0           NaN                                               NaN   
1           NaN                                               NaN   
2        FRS Id                                     Facility Name   
3  110070930991                              Atmos Pipeline Texas   
4  110062051880  Consumers Energy Company - Transmission Pipeline   

                    Unnamed: 3     Unnamed: 4      Unnamed: 5  \
0                          NaN            NaN             NaN   
1                          NaN            NaN             NaN   
2  State where Emissions Occur  Reported City  Reported State   
3                        Texas         Dallas              TX   
4                     Michigan        Jackson              MI   

          Unnamed: 6        Unnamed: 7       Unnamed: 8         Unnamed: 9  \
0                NaN               NaN              NaN                NaN   
1                NaN               NaN              NaN                NaN   
2  Reported Zip Code  Reported Address  Reported County  Reported Latitude   
3              75240  5430 LBJ Freeway    DALLAS COUNTY           32.92548   
4              49201  One Energy Plaza   JACKSON COUNTY           42.24668   

          Unnamed: 10         Unnamed: 11               Unnamed: 12  \
0                 NaN                 NaN                       NaN   
1                 NaN                 NaN                       NaN   
2  Reported Longitude  Primary NAICS Code  Industry Type (subparts)   
3           -96.81613              486210                   W-TRANS   
4           -84.40311              486210                   W-TRANS   

                                         Unnamed: 13  \
0                                                NaN   
1                                                NaN   
2  Total reported direct emissions from Transmiss...   
3                                       21465.090362   
4                                      139434.403542   

                                         Unnamed: 14               Unnamed: 15  
0                                                NaN                       NaN  
1  Total Direct Emissions from Transmission Pipel...                       NaN  
2                      CO2 emissions (non-biogenic)   Methane (CH4) emissions   
3                                          55.859402               21409.23096  
4                                         160.653541                 139273.75

In [17]:
ldc_data.head()

Summary data collected by the Greenhouse Gas Reporting Program for 2022  \
0  This data was reported to EPA by facilities as...                        
1  All emissions data is presented in units of me...                        
2                                        Facility Id                        
3                                            1008026                        
4                                            1004034                        

     Unnamed: 1       Unnamed: 2                   Unnamed: 3     Unnamed: 4  \
0           NaN              NaN                          NaN            NaN   
1           NaN              NaN                          NaN            NaN   
2        FRS Id    Facility Name  State where Emissions Occur  Reported City   
3  110071160130  Ameren Illinois                           IL         Peoria   
4  110071160131  Ameren Missouri                           MO      St. Louis   

       Unnamed: 5         Unnamed: 6            Unnamed: 7       Unnamed: 8  \
0             NaN                NaN                   NaN              NaN   
1             NaN                NaN                   NaN              NaN   
2  Reported State  Reported Zip Code      Reported Address  Reported County   
3              IL              61602    300 Liberty Street    PEORIA COUNTY   
4              MO              63103  1901 Chouteau Avenue   ST. LOUIS CITY   

          Unnamed: 9         Unnamed: 10         Unnamed: 11  \
0                NaN                 NaN                 NaN   
1                NaN                 NaN                 NaN   
2  Reported Latitude  Reported Longitude  Primary NAICS Code   
3           40.69045            -89.5924              221210   
4          38.620667          -90.211086              221210   

                Unnamed: 12  \
0                       NaN   
1                       NaN   
2  Industry Type (subparts)   
3              NN-LDC,W-LDC   
4              NN-LDC,W-LDC   

                                         Unnamed: 13  \
0                                                NaN   
1                                                NaN   
2  Total reported direct emissions from Local Dis...   
3                                          82377.892   
4                                            15062.1   

                                         Unnamed: 14  \
0                                                NaN   
1  Total Direct Emissions from Local Distribution...   
2                      CO2 emissions (non-biogenic)    
3                                             2164.2   
4                                               18.1   

                Unnamed: 15                     Unnamed: 16  \
0                       NaN                             NaN   
1                       NaN                             NaN   
2  Methane (CH4) emissions   Nitrous Oxide (N2O) emissions    
3                   80212.5                           1.192   
4                     15044                             NaN   

                                         Unnamed: 17  
0                                                NaN  
1                                                NaN  
2  Does the facility employ continuous emissions ...  
3                                                  N  
4                                                  N

In [18]:
sf6_elec_data.head()

Summary data collected by the Greenhouse Gas Reporting Program for 2022  \
0  This data was reported to EPA by facilities as...                        
1  All emissions data is presented in units of me...                        
2                                        Facility Id                        
3                                            1003708                        
4                                            1010811                        

     Unnamed: 1                   Unnamed: 2     Unnamed: 3      Unnamed: 4  \
0           NaN                          NaN            NaN             NaN   
1           NaN                          NaN            NaN             NaN   
2        FRS Id                Facility Name  Reported City  Reported State   
3  110000591930  AEP Corporate SF6 Emissions       Columbus              OH   
4  110010724216     Ameren Illinois Electric         Peoria              IL   

          Unnamed: 5          Unnamed: 6         Unnamed: 7  \
0                NaN                 NaN                NaN   
1                NaN                 NaN                NaN   
2  Reported Zip Code    Reported Address    Reported County   
3              43215   1 Riverside Plaza  WASHINGTON COUNTY   
4              61602  300 Liberty Street      PEORIA COUNTY   

          Unnamed: 8          Unnamed: 9         Unnamed: 10  \
0                NaN                 NaN                 NaN   
1                NaN                 NaN                 NaN   
2  Reported Latitude  Reported Longitude  Primary NAICS Code   
3          39.588493          -81.682523              221112   
4           40.69048           -89.59244              221122   

                Unnamed: 11  \
0                       NaN   
1                       NaN   
2  Industry Type (subparts)   
3                        DD   
4                        DD   

                                         Unnamed: 12     Unnamed: 13  \
0                                                NaN             NaN   
1                                                NaN             NaN   
2  Total reported direct emissions from Electrica...  SF6 emissions    
3                                         76825.8312      76825.8312   
4                                          7667.5944       7667.5944   

                                         Unnamed: 14  
0                                                NaN  
1                                                NaN  
2  Does the facility employ continuous emissions ...  
3                                                  N  
4                                                  N

In [19]:
indust_type_data.head()

Unnamed: 0  Unnamed: 1      Unnamed: 2              Unnamed: 3  \
0         NaN         NaN  Subpart Letter        Name of industry   
1         NaN         NaN               C   Stationary Combustion   
2         NaN         NaN               D  Electricity Generation   
3         NaN         NaN               E  Adipic Acid Production   
4         NaN         NaN               F     Aluminum Production   

       Unnamed: 4  
0   Facility Type  
1  Direct Emitter  
2  Direct Emitter  
3  Direct Emitter  
4  Direct Emitter

***changing column headers***

In [3]:
# the columns headers are unnamed for each of the datasets, with the actual headers as row 0 
# need to assign row 2 as column header for each dataset 
dir_emit_data.columns = dir_emit_data.iloc[2]
dir_emit_data.head()

2                                        Facility Id        FRS Id  \
0  This data was reported to EPA by facilities as...           NaN   
1  All emissions data is presented in units of me...           NaN   
2                                        Facility Id        FRS Id   
3                                            1002685  110071162081   
4                                            1004377  110043803578   

2                   Facility Name      City  State  Zip Code  \
0                             NaN       NaN    NaN       NaN   
1                             NaN       NaN    NaN       NaN   
2                   Facility Name      City  State  Zip Code   
3                        AC 857 A  Offshore     TX     00000   
4  121 REGIONAL DISPOSAL FACILITY   MELISSA     TX     75454   

2                   Address         County   Latitude  Longitude  ...  \
0                       NaN            NaN        NaN        NaN  ...   
1                       NaN            NaN        NaN        NaN  ...   
2                   Address         County   Latitude  Longitude  ...   
3                       NaN            NaN  26.295556 -94.897778  ...   
4  3820 SAM RAYBURN HIGHWAY  COLLIN COUNTY   33.29857  -96.53586  ...   

2  Titanium Dioxide Production  Underground Coal Mines  Zinc Production  \
0                          NaN                     NaN              NaN   
1                          NaN                     NaN              NaN   
2  Titanium Dioxide Production  Underground Coal Mines  Zinc Production   
3                          NaN                     NaN              NaN   
4                          NaN                     NaN              NaN   

2  Municipal Landfills  Industrial Wastewater Treatment  \
0                  NaN                              NaN   
1                  NaN                              NaN   
2  Municipal Landfills  Industrial Wastewater Treatment   
3                  NaN                              NaN   
4            325416.25                              NaN   

2 Manufacture of Electric Transmission and Distribution Equipment  \
0                                                NaN                
1                                                NaN                
2  Manufacture of Electric Transmission and Distr...                
3                                                NaN                
4                                                NaN                

2  Industrial Waste Landfills  \
0                         NaN   
1                         NaN   
2  Industrial Waste Landfills   
3                         NaN   
4                         NaN   

2 Is some CO2 collected on-site and used to manufacture other products and therefore not emitted from the affected manufacturing process unit(s)? (as reported under Subpart G or S)  \
0                                                NaN                                                                                                                                   
1                                                NaN                                                                                                                                   
2  Is some CO2 collected on-site and used to manu...                                                                                                                                   
3                                                  N                                                                                                                                   
4                                                  N                                                                                                                                   

2 Is some CO2 reported as emissions from the affected manufacturing process unit(s) under Subpart AA, G or P collected and transferred off-site or injected (as reported under Subpart PP)?  \
0                                      

In [4]:
# now dropping the first three rows since the first two were useless information & the third was the row I used for the headers
dir_emit_data = dir_emit_data.drop([0, 1, 2])
dir_emit_data.head()
# looks good!

2 Facility Id        FRS Id                   Facility Name         City  \
3     1002685  110071162081                        AC 857 A     Offshore   
4     1004377  110043803578  121 REGIONAL DISPOSAL FACILITY      MELISSA   
5     1010040  110071159872               15-18565/15-18662       Hazard   
6     1000112  110071159492                    23rd and 3rd     BROOKLYN   
7     1003742  110038508335            31st Street Landfill  WESTCHESTER   

2 State Zip Code                   Address         County   Latitude  \
3    TX    00000                       NaN            NaN  26.295556   
4    TX    75454  3820 SAM RAYBURN HIGHWAY  COLLIN COUNTY   33.29857   
5    KY    40701           1021 Tori Drive   PERRY COUNTY  37.274127   
6    NY    11232            730 3rd Avenue          Kings     40.663   
7    IL    60154           11700 W 31ST ST    COOK COUNTY  41.834962   

2  Longitude  ... Titanium Dioxide Production Underground Coal Mines  \
3 -94.897778  ...                         NaN                    NaN   
4  -96.53586  ...                         NaN                    NaN   
5 -83.239034  ...                         NaN               51538.75   
6        -74  ...                         NaN                    NaN   
7 -87.916392  ...                         NaN                    NaN   

2 Zinc Production Municipal Landfills Industrial Wastewater Treatment  \
3             NaN                 NaN                             NaN   
4             NaN           325416.25                             NaN   
5             NaN                 NaN                             NaN   
6             NaN                 NaN                             NaN   
7             NaN           105999.75                             NaN   

2 Manufacture of Electric Transmission and Distribution Equipment  \
3                                                NaN                
4                                                NaN                
5                                                NaN                
6                                                NaN                
7                                                NaN                

2 Industrial Waste Landfills  \
3                        NaN   
4                        NaN   
5                        NaN   
6                        NaN   
7                        NaN   

2 Is some CO2 collected on-site and used to manufacture other products and therefore not emitted from the affected manufacturing process unit(s)? (as reported under Subpart G or S)  \
3                                                  N                                                                                                                                   
4                                                  N                                                                                                                                   
5                                                  N                                                                                                                                   
6                                                  N                                                                                                                                   
7                                                  N                                                                                                                                   

2 Is some CO2 reported as emissions from the affected manufacturing process unit(s) under Subpart AA, G or P collected and transferred off-site or injected (as reported under Subpart PP)?  \
3                                                  N                                                                                                                                          
4                                                  N                                                                                           

In [5]:
# now doing the above process for each dataframe to get the right headers for each 

# starting with onshore gas & oil df 
# changing headers
onshore_gas_oil_data.columns = onshore_gas_oil_data.iloc[2]

In [6]:
# now dropping the relevant onshore gas & oil rows 
onshore_gas_oil_data = onshore_gas_oil_data.drop([0, 1, 2])
onshore_gas_oil_data.head()
# looks good! 

2 Facility Id        FRS Id  \
3     1000355           NaN   
4     1009238           NaN   
5     1009170           NaN   
6     1009265  110071162284   
7     1008287           NaN   

2                                      Facility Name  \
3             (540) Civitas Resources - Denver Basin   
4                           220 Gulf Coast Basin DEC   
5  260 East Texas Basin - BP America Production C...   
6                                 360 Anadarko Basin   
7                             360 Anadarko Basin DEC   

2                            Basin  Reported City Reported State  \
3               540 - Denver Basin         Denver             CO   
4  220 - Gulf Coast Basin (LA, TX)  Oklahoma City             OK   
5           260 - East Texas Basin        Houston             TX   
6             360 - Anadarko Basin  Oklahoma City             OK   
7             360 - Anadarko Basin  Oklahoma City             OK   

2 Reported Zip Code                   Reported Address Reported County  \
3             80202            555 17th St. Suite 3700             NaN   
4             73102              333 West Sheridan Ave             NaN   
5             77079            501 Westlake Park Blvd.             NaN   
6             73142  4700 Gaillardia Parkway, Ste. 200             NaN   
7             73102              333 West Sheridan Ave             NaN   

2 Reported Latitude Reported Longitude Primary NAICS Code  \
3          39.74431         -104.98858             211120   
4         35.466697          -97.51453             211120   
5          29.78037           -95.6295             211130   
6          35.56719          -97.49827             211120   
7         35.466697          -97.51453             211120   

2 Industry Type (subparts)  \
3                   W-ONSH   
4                   W-ONSH   
5                   W-ONSH   
6                   W-ONSH   
7                   W-ONSH   

2 Total reported emissions from Onshore Oil & Gas Production   \
3                                        1171208.898            
4                                        1173336.492            
5                                          15645.134            
6                                         147115.318            
7                                         349859.078            

2 CO2 emissions (non-biogenic)  Methane (CH4) emissions   \
3                      638405.5                 532296.5   
4                      913156.4                259679.75   
5                        3719.5                 11923.25   
6                       97548.9                 49472.25   
7                      262643.6                 86981.25   

2 Nitrous Oxide (N2O) emissions   \
3                        506.898   
4                        500.342   
5                          2.384   
6                         94.168   
7                        234.228   

2 Does the facility employ continuous emissions monitoring?   
3                                                  N          
4                                                  N          
5                                                  N          
6                                                  N          
7                                                  N

In [7]:
# now doing the gathering & boosting df
# changing headers...
gat_boost_data.columns = gat_boost_data.iloc[2]

In [8]:
# dropping relevant gathering & boosting rows 
gat_boost_data = gat_boost_data.drop([0, 1, 2])
gat_boost_data.head() 
# looks good! 

2 Facility Id        FRS Id                                Facility Name  \
3     1014558           NaN  (540) Civitas Resources - GB - Denver Basin   
4     1012507  110070082329                         220 Gulf Coast Basin   
5     1012472  110070082329                             345 Arkoma Basin   
6     1012466  110070082329               350 South Oklahoma Folded Belt   
7     1012522  110071099264                           360 Anadarko Basin   

2                             Basin Reported City Reported State  \
3                540 - Denver Basin        Denver             CO   
4   220 - Gulf Coast Basin (LA, TX)       Houston             TX   
5                345 - Arkoma Basin       Houston             TX   
6  350 - South Oklahoma Folded Belt       Houston             TX   
7              360 - Anadarko Basin       Houston             TX   

2 Reported Zip Code           Reported Address Reported County  \
3             80202    555 17th St. Suite 3700             NaN   
4             77002  811 Louisiana, Suite 2100   HARRIS COUNTY   
5             77002  811 Louisiana, Suite 2100   HARRIS COUNTY   
6             77002  811 Louisiana, Suite 2100   HARRIS COUNTY   
7             77002   811 Louisiana Suite 2100   HARRIS COUNTY   

2 Reported Latitude Reported Longitude Primary NAICS Code  \
3         39.745822        -104.989243             211120   
4          29.75993         -95.366413             211130   
5          29.75993         -95.366413             211130   
6          29.75993         -95.366413             211130   
7          29.75993         -95.366413             211130   

2 Industry Type (subparts) Total reported emissions from Gathering & Boosting  \
3                     W-GB                                          79593.288   
4                     W-GB                                      228165.616365   
5                     W-GB                                        66774.54771   
6                     W-GB                                       90945.590167   
7                     W-GB                                      254529.554488   

2 CO2 emissions (non-biogenic)  Methane (CH4) emissions   \
3                       73618.7                     5943   
4                 201806.334775              26249.79805   
5                  62693.583621              4048.444627   
6                   78600.32211             12302.231832   
7                 206475.504843             47939.336617   

2 Nitrous Oxide (N2O) emissions   
3                         31.588  
4                      109.48354  
5                      32.519462  
6                      43.036225  
7                     114.713028

In [9]:
# now doing tranmission pipe df 
tran_pipe_data.columns = tran_pipe_data.iloc[2]

In [10]:
# dropping rows....
tran_pipe_data = tran_pipe_data.drop([0, 1, 2])
tran_pipe_data.head()
# looks good 

2 Facility Id        FRS Id  \
3     1013856  110070930991   
4     1012634  110062051880   
5     1013773  110004647142   
6     1013663  110070716075   
7     1013663  110070716075   

2                                      Facility Name  \
3                               Atmos Pipeline Texas   
4   Consumers Energy Company - Transmission Pipeline   
5  Dominion Energy Ohio - Transmission Pipelines ...   
6                                       EGT Arkansas   
7                                       EGT Arkansas   

2 State where Emissions Occur  Reported City Reported State Reported Zip Code  \
3                       Texas         Dallas             TX             75240   
4                    Michigan        Jackson             MI             49201   
5                        Ohio      Cleveland             OH             44103   
6                    Arkansas  Oklahoma City             OK             73100   
7                       Texas  Oklahoma City             OK             73100   

2             Reported Address  Reported County Reported Latitude  \
3             5430 LBJ Freeway    DALLAS COUNTY          32.92548   
4             One Energy Plaza   JACKSON COUNTY          42.24668   
5        1201 East 55th Street  CUYAHOGA COUNTY          41.52712   
6  499 W. Sheridan, Suite 1500  OKLAHOMA COUNTY          35.46646   
7  499 W. Sheridan, Suite 1500  OKLAHOMA COUNTY          35.46646   

2 Reported Longitude Primary NAICS Code Industry Type (subparts)  \
3          -96.81613             486210                  W-TRANS   
4          -84.40311             486210                  W-TRANS   
5            -81.652             221210                  W-TRANS   
6          -97.52124             486210                  W-TRANS   
7          -97.52124             486210                  W-TRANS   

2 Total reported direct emissions from Transmission Pipelines  \
3                                       21465.090362            
4                                      139434.403542            
5                                        5197.776773            
6                                         9278.99582            
7                                          696.25597            

2 CO2 emissions (non-biogenic)  Methane (CH4) emissions   
3                     55.859402              21409.23096  
4                    160.653541                139273.75  
5                      5.988773                 5191.788  
6                      10.69107               9268.30475  
7                       0.80222                695.45375

In [11]:
# now doing local distribution companies df 
ldc_data.columns = ldc_data.iloc[2]

In [12]:
# dropping rows here....
ldc_data = ldc_data.drop([0, 1, 2])
ldc_data.head()

2 Facility Id        FRS Id                        Facility Name  \
3     1008026  110071160130                      Ameren Illinois   
4     1004034  110071160131                      Ameren Missouri   
5     1007872  110071161198            Atlanta Gas Light Company   
6     1004794  110070081906  Atmos Energy Corporation - Colorado   
7     1001388  110071162238    Atmos Energy Corporation - Kansas   

2 State where Emissions Occur Reported City Reported State Reported Zip Code  \
3                          IL        Peoria             IL             61602   
4                          MO     St. Louis             MO             63103   
5                          GA       Atlanta             GA             30309   
6                          CO        Dallas             TX             75240   
7                          KS        Dallas             TX             75240   

2      Reported Address Reported County Reported Latitude Reported Longitude  \
3    300 Liberty Street   PEORIA COUNTY          40.69045           -89.5924   
4  1901 Chouteau Avenue  ST. LOUIS CITY         38.620667         -90.211086   
5   Ten Peachtree Place   FULTON COUNTY         33.797114         -84.380489   
6      5430 LBJ Freeway   DALLAS COUNTY         32.925488         -96.816137   
7      5430 LBJ Freeway   DALLAS COUNTY         32.925488         -96.816137   

2 Primary NAICS Code Industry Type (subparts)  \
3             221210             NN-LDC,W-LDC   
4             221210             NN-LDC,W-LDC   
5             221210             NN-LDC,W-LDC   
6             221210             NN-LDC,W-LDC   
7             221210             NN-LDC,W-LDC   

2 Total reported direct emissions from Local Distribution Companies  \
3                                          82377.892                  
4                                            15062.1                  
5                                          175745.35                  
6                                            23799.7                  
7                                           23232.25                  

2 CO2 emissions (non-biogenic)  Methane (CH4) emissions   \
3                        2164.2                  80212.5   
4                          18.1                    15044   
5                       11749.6                163995.75   
6                          28.7                    23771   
7                            28                 23204.25   

2 Nitrous Oxide (N2O) emissions   \
3                          1.192   
4                            NaN   
5                            NaN   
6                            NaN   
7                            NaN   

2 Does the facility employ continuous emissions monitoring?   
3                                                  N          
4                                                  N          
5                                                  N          
6                                                  N          
7                                                  N

In [13]:
# now doing SF6 df
sf6_elec_data.columns = sf6_elec_data.iloc[2]

In [14]:
# dropping...
sf6_elec_data = sf6_elec_data.drop([0, 1, 2])
sf6_elec_data.head()

2 Facility Id        FRS Id                         Facility Name  \
3     1003708  110000591930           AEP Corporate SF6 Emissions   
4     1010811  110010724216              Ameren Illinois Electric   
5     1010812  110054133479              Ameren Missouri Electric   
6     1010076  110066446643  Anaheim Electric Distribution System   
7     1009780  110011802148        Arizona Public Service Company   

2 Reported City Reported State Reported Zip Code  \
3      Columbus             OH             43215   
4        Peoria             IL             61602   
5     St. Louis             MO             63103   
6       Anaheim             CA             92805   
7       Phoenix             AZ             85021   

2               Reported Address    Reported County Reported Latitude  \
3              1 Riverside Plaza  WASHINGTON COUNTY         39.588493   
4             300 Liberty Street      PEORIA COUNTY          40.69048   
5           1901 Chouteau Avenue     ST. LOUIS CITY         38.620785   
6  201 S. Anaheim Blvd., MS-1101      ORANGE COUNTY          33.83458   
7            2124 W. Cheryl Ave.    MARICOPA COUNTY         33.578425   

2 Reported Longitude Primary NAICS Code Industry Type (subparts)  \
3         -81.682523             221112                       DD   
4          -89.59244             221122                       DD   
5         -90.211635             221122                       DD   
6         -117.91401             221122                       DD   
7        -112.104408             221121                       DD   

2 Total reported direct emissions from Electrical Equipment Use  \
3                                         76825.8312              
4                                          7667.5944              
5                                           13142.49              
6                                          2226.6024              
7                                         10761.0528              

2 SF6 emissions  Does the facility employ continuous emissions monitoring?   
3     76825.8312                                                  N          
4      7667.5944                                                  N          
5       13142.49                                                  N          
6      2226.6024                                                  N          
7     10761.0528                                                  N

In [15]:
# now doing the industry type df 
# this one is slightly different because there is only one useless row, so the iloc will be a different value 
# also the first two columns are useless, so dropping that first here
indust_type_data = indust_type_data.drop(columns = ['Unnamed: 0', 'Unnamed: 1'])
indust_type_data.head()
# looks good! 


Unnamed: 2              Unnamed: 3      Unnamed: 4
0  Subpart Letter        Name of industry   Facility Type
1               C   Stationary Combustion  Direct Emitter
2               D  Electricity Generation  Direct Emitter
3               E  Adipic Acid Production  Direct Emitter
4               F     Aluminum Production  Direct Emitter

In [16]:
# now doing iloc function: 
indust_type_data.columns = indust_type_data.iloc[0]
indust_type_data.head()

0  Subpart Letter        Name of industry   Facility Type
0  Subpart Letter        Name of industry   Facility Type
1               C   Stationary Combustion  Direct Emitter
2               D  Electricity Generation  Direct Emitter
3               E  Adipic Acid Production  Direct Emitter
4               F     Aluminum Production  Direct Emitter

In [17]:
# dropping irrelevant row
indust_type_data = indust_type_data.drop([0])
indust_type_data.head()

0 Subpart Letter        Name of industry   Facility Type
1              C   Stationary Combustion  Direct Emitter
2              D  Electricity Generation  Direct Emitter
3              E  Adipic Acid Production  Direct Emitter
4              F     Aluminum Production  Direct Emitter
5              G   Ammonia Manufacturing  Direct Emitter

***dropping and renaming columns in dfs***

In [18]:
# starting with direct emissions df
dir_emit_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6491 entries, 3 to 6493
Data columns (total 66 columns):
 #   Column                                                                                                                                                                                     Non-Null Count  Dtype 
---  ------                                                                                                                                                                                     --------------  ----- 
 0   Facility Id                                                                                                                                                                                6491 non-null   object
 1   FRS Id                                                                                                                                                                                     6325 non-null   object
 2   Facility Name               

In [19]:
# now going to drop the columns that I will not be using & rename columns so that joining will be cleaner 
# rename...CO2 emissions (non-biogenic) to CO2 emissions; Total reported direct emission to Total reported emissions
# dropping...columns 5:10, 12, 17:18, 20:65

dir_emit_data = dir_emit_data.drop(columns=dir_emit_data.columns[[5, 6, 7, 8, 9, 10, 12, 17, 18] + list(range(20, 66))])

In [20]:
# stripping any whitespace to make sure the columns are renamed 
dir_emit_data.columns = dir_emit_data.columns.str.strip()
final_dir_emit_data = dir_emit_data.rename(columns = {'CO2 emissions (non-biogenic)': 'CO2 emissions', 'Total reported direct emissions': 'Total reported emissions'})

In [21]:
# taking a look to make sure everything is good in df
final_dir_emit_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6491 entries, 3 to 6493
Data columns (total 11 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   Facility Id                    6491 non-null   object
 1   FRS Id                         6325 non-null   object
 2   Facility Name                  6491 non-null   object
 3   City                           6491 non-null   object
 4   State                          6491 non-null   object
 5   Industry Type (subparts)       6491 non-null   object
 6   Total reported emissions       6491 non-null   object
 7   CO2 emissions                  5841 non-null   object
 8   Methane (CH4) emissions        6383 non-null   object
 9   Nitrous Oxide (N2O) emissions  5447 non-null   object
 10  SF6 emissions                  59 non-null     object
dtypes: object(11)
memory usage: 557.9+ KB


In [21]:
# onto the next df...gathering & boosting df!
# looking at the info first 
gat_boost_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 3 to 352
Data columns (total 17 columns):
 #   Column                                              Non-Null Count  Dtype 
---  ------                                              --------------  ----- 
 0   Facility Id                                         350 non-null    object
 1   FRS Id                                              310 non-null    object
 2   Facility Name                                       350 non-null    object
 3   Basin                                               346 non-null    object
 4   Reported City                                       350 non-null    object
 5   Reported State                                      350 non-null    object
 6   Reported Zip Code                                   350 non-null    object
 7   Reported Address                                    334 non-null    object
 8   Reported County                                     262 non-null    object
 9   Reported L

In [22]:
# dropping....3, 6:11 
# renaming...Reported City to City; Reported State to State; Total reported emissions from Gathering & Boosting to Total reported emissions; CO2 emissions (non-biogenic) to CO2 emissions

gat_boost_data = gat_boost_data.drop(columns = gat_boost_data.columns[[3] + list(range(6, 12))])

In [23]:
# renaming time 
#first, getting rid of whitespace
gat_boost_data.columns = gat_boost_data.columns.str.strip()
final_gat_boost_data = gat_boost_data.rename(columns = {'Reported City' : 'City', 'Reported State' : 'State', 'Total reported emissions from Gathering & Boosting' : 'Total reported emissions', 'CO2 emissions (non-biogenic)' : 'CO2 emissions'})

In [24]:
# checking to make sure 
final_gat_boost_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 3 to 352
Data columns (total 10 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   Facility Id                    350 non-null    object
 1   FRS Id                         310 non-null    object
 2   Facility Name                  350 non-null    object
 3   City                           350 non-null    object
 4   State                          350 non-null    object
 5   Industry Type (subparts)       350 non-null    object
 6   Total reported emissions       350 non-null    object
 7   CO2 emissions                  349 non-null    object
 8   Methane (CH4) emissions        349 non-null    object
 9   Nitrous Oxide (N2O) emissions  340 non-null    object
dtypes: object(10)
memory usage: 27.5+ KB


In [24]:
# doing ldc df
ldc_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 161 entries, 3 to 163
Data columns (total 18 columns):
 #   Column                                                             Non-Null Count  Dtype 
---  ------                                                             --------------  ----- 
 0   Facility Id                                                        161 non-null    object
 1   FRS Id                                                             153 non-null    object
 2   Facility Name                                                      161 non-null    object
 3   State where Emissions Occur                                        161 non-null    object
 4   Reported City                                                      161 non-null    object
 5   Reported State                                                     161 non-null    object
 6   Reported Zip Code                                                  161 non-null    object
 7   Reported Address                   

In [25]:
# dropping...3, 6:11, 17
# renaming...Reported City to City; Reported State to State, Total reported direct emissions from Local Distribution Companies to Total reported emissions; CO2 emisisons (non-biogenic) to CO2 emissions

ldc_data = ldc_data.drop(columns = ldc_data.columns[[3] + list(range(6, 12)) + [17]])

In [26]:
# renaming time
# stripping whitespace first 
ldc_data.columns = ldc_data.columns.str.strip()
final_ldc_data = ldc_data.rename(columns = {'Reported City' : 'City', 'Reported State' : 'State', 'Total reported direct emissions from Local Distribution Companies' : 'Total reported emissions', 'CO2 emissions (non-biogenic)' : 'CO2 emissions'})

In [27]:
# making sure everything looks good
final_ldc_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 161 entries, 3 to 163
Data columns (total 10 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   Facility Id                    161 non-null    object
 1   FRS Id                         153 non-null    object
 2   Facility Name                  161 non-null    object
 3   City                           161 non-null    object
 4   State                          161 non-null    object
 5   Industry Type (subparts)       161 non-null    object
 6   Total reported emissions       161 non-null    object
 7   CO2 emissions                  161 non-null    object
 8   Methane (CH4) emissions        161 non-null    object
 9   Nitrous Oxide (N2O) emissions  27 non-null     object
dtypes: object(10)
memory usage: 12.7+ KB


In [38]:
# doing onshore gas oil df 
# looking at info first 
onshore_gas_oil_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 451 entries, 3 to 453
Data columns (total 18 columns):
 #   Column                                                       Non-Null Count  Dtype 
---  ------                                                       --------------  ----- 
 0   Facility Id                                                  451 non-null    object
 1   FRS Id                                                       95 non-null     object
 2   Facility Name                                                451 non-null    object
 3   Basin                                                        444 non-null    object
 4   Reported City                                                451 non-null    object
 5   Reported State                                               451 non-null    object
 6   Reported Zip Code                                            451 non-null    object
 7   Reported Address                                             445 non-null    object
 8   

In [27]:
# dropping...3, 6:11, 17
# renaming.. Reported City to City; Reported State to State; Total reported emissions from Onshore Oil & Gas Production to Total reported emissions; CO2 emissions (non-biogenic) to CO2 emisisons

# dropping: 
onshore_gas_oil_data = onshore_gas_oil_data.drop(columns = onshore_gas_oil_data.columns[[3] + list(range(6, 12)) + [17]])

In [28]:
# renaming part: 
# stripping whitetrail first 
onshore_gas_oil_data.columns = onshore_gas_oil_data.columns.str.strip()
final_onshore_data = onshore_gas_oil_data.rename(columns = {'Reported City' : 'City', 'Reported State' : 'State', 'Total reported emissions from Onshore Oil & Gas Production' : 'Total reported emissions', 'CO2 emissions (non-biogenic)' : 'CO2 emissions'})

In [29]:
# making sure everything is correct 
final_onshore_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 451 entries, 3 to 453
Data columns (total 10 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   Facility Id                    451 non-null    object
 1   FRS Id                         95 non-null     object
 2   Facility Name                  451 non-null    object
 3   City                           451 non-null    object
 4   State                          451 non-null    object
 5   Industry Type (subparts)       451 non-null    object
 6   Total reported emissions       451 non-null    object
 7   CO2 emissions                  451 non-null    object
 8   Methane (CH4) emissions        450 non-null    object
 9   Nitrous Oxide (N2O) emissions  433 non-null    object
dtypes: object(10)
memory usage: 35.4+ KB


In [30]:
# doing sf6_elec_data df now!
# looking at info...
sf6_elec_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 93 entries, 3 to 95
Data columns (total 15 columns):
 #   Column                                                         Non-Null Count  Dtype 
---  ------                                                         --------------  ----- 
 0   Facility Id                                                    93 non-null     object
 1   FRS Id                                                         89 non-null     object
 2   Facility Name                                                  93 non-null     object
 3   Reported City                                                  93 non-null     object
 4   Reported State                                                 93 non-null     object
 5   Reported Zip Code                                              93 non-null     object
 6   Reported Address                                               86 non-null     object
 7   Reported County                                                89 non-nul

In [31]:
# dropping...5-10, 14
# renaming.. Reported City to City; Reported State to State; Total reported direct emissions from Electrical Equipment Use to Total reported emissions

# dropping: 
sf6_elec_data = sf6_elec_data.drop(columns = sf6_elec_data.columns[list(range(5, 11)) + [14]])

In [32]:
# renaming part! 
# stripping whitetrail first 
sf6_elec_data.columns = sf6_elec_data.columns.str.strip()
final_elec_data = sf6_elec_data.rename(columns = {'Reported City' : 'City', 'Reported State' : 'State', 'Total reported direct emissions from Electrical Equipment Use' : 'Total reported emissions'})

In [33]:
# making sure it's all goood
final_elec_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 93 entries, 3 to 95
Data columns (total 8 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Facility Id               93 non-null     object
 1   FRS Id                    89 non-null     object
 2   Facility Name             93 non-null     object
 3   City                      93 non-null     object
 4   State                     93 non-null     object
 5   Industry Type (subparts)  93 non-null     object
 6   Total reported emissions  93 non-null     object
 7   SF6 emissions             75 non-null     object
dtypes: object(8)
memory usage: 5.9+ KB


In [34]:
# now piplines df!
tran_pipe_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 158 entries, 3 to 160
Data columns (total 16 columns):
 #   Column                                                       Non-Null Count  Dtype 
---  ------                                                       --------------  ----- 
 0   Facility Id                                                  158 non-null    object
 1   FRS Id                                                       152 non-null    object
 2   Facility Name                                                158 non-null    object
 3   State where Emissions Occur                                  158 non-null    object
 4   Reported City                                                158 non-null    object
 5   Reported State                                               158 non-null    object
 6   Reported Zip Code                                            158 non-null    object
 7   Reported Address                                             158 non-null    object
 8   

In [35]:
# dropping...3, 6-11
# renaming.. Reported City to City; Reported State to State; Total reported direct emissions from Transmission Pipelines to Total reported emissions; CO2 emissions (non-biogenic) to CO2 emisisons

# dropping: 
tran_pipe_data = tran_pipe_data.drop(columns = tran_pipe_data.columns[[3] + list(range(6, 12))])

In [36]:
# renaming now 
# stripping whitetrail:
tran_pipe_data.columns = tran_pipe_data.columns.str.strip()
final_pipe_data = tran_pipe_data.rename(columns = {'Reported City' : 'City', 'Reported State' : 'State', 'Total reported direct emissions from Transmission Pipelines' : 'Total reported emissions', 'CO2 emissions (non-biogenic)' : 'CO2 emissions'})

In [37]:
# making sure everything is good 
final_pipe_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 158 entries, 3 to 160
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Facility Id               158 non-null    object
 1   FRS Id                    152 non-null    object
 2   Facility Name             158 non-null    object
 3   City                      158 non-null    object
 4   State                     158 non-null    object
 5   Industry Type (subparts)  158 non-null    object
 6   Total reported emissions  158 non-null    object
 7   CO2 emissions             155 non-null    object
 8   Methane (CH4) emissions   158 non-null    object
dtypes: object(9)
memory usage: 11.2+ KB


***joining dataframes together to make one big df***

In [38]:
# now joining the dfs together to make one big df 
# joining the emitting dfs by 
# joining the industry type data by subpart letter (in indust_type_data) to industry type in the emitting dfs

# joining the emititng dfs first 
frames = [final_dir_emit_data, final_elec_data, final_gat_boost_data, final_ldc_data, final_onshore_data, final_pipe_data]

data = pd.concat(frames, join = "outer") 
data.head()

2 Facility Id        FRS Id                   Facility Name         City  \
3     1002685  110071162081                        AC 857 A     Offshore   
4     1004377  110043803578  121 REGIONAL DISPOSAL FACILITY      MELISSA   
5     1010040  110071159872               15-18565/15-18662       Hazard   
6     1000112  110071159492                    23rd and 3rd     BROOKLYN   
7     1003742  110038508335            31st Street Landfill  WESTCHESTER   

2 State Industry Type (subparts) Total reported emissions CO2 emissions  \
3    TX                C,W-OFFSH               205454.158      200728.2   
4    TX                       HH                325416.25           NaN   
5    KY                       FF                 51538.75           NaN   
6    NY                      C,D                68839.396       68769.8   
7    IL                     C,HH               106294.198         293.9   

2 Methane (CH4) emissions Nitrous Oxide (N2O) emissions SF6 emissions  
3                 4563.25                       162.708           NaN  
4               325416.25                           NaN           NaN  
5                51538.75                           NaN           NaN  
6                   31.75                        37.846           NaN  
7                  106000                         0.298           NaN

In [39]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7704 entries, 3 to 160
Data columns (total 11 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   Facility Id                    7704 non-null   object
 1   FRS Id                         7124 non-null   object
 2   Facility Name                  7704 non-null   object
 3   City                           7704 non-null   object
 4   State                          7704 non-null   object
 5   Industry Type (subparts)       7704 non-null   object
 6   Total reported emissions       7704 non-null   object
 7   CO2 emissions                  6957 non-null   object
 8   Methane (CH4) emissions        7501 non-null   object
 9   Nitrous Oxide (N2O) emissions  6247 non-null   object
 10  SF6 emissions                  134 non-null    object
dtypes: object(11)
memory usage: 722.2+ KB


In [41]:
# now going to join the 'data' df to the industry type df 
# I am using the partial string joining method
    # using this method because the data df values in 'industry type (subparts) column doesn't perfectly align with the 'subpart letter' column in indust_type_data
        # the industry type (subparts) column lists multiple subparts, which is why they don't perfectly align 

data = pd.merge(data[data['Industry Type (subparts)'].str.contains('|'.join(indust_type_data['Subpart Letter']))],
                indust_type_data,
                left_on='Industry Type (subparts)',
                right_on='Subpart Letter',           
                how='inner')

data.head()

Facility Id        FRS Id  \
0     1004377  110043803578   
1     1004680  110013945062   
2     1012058  110071161547   
3     1006529  110013921435   
4     1001991  110071160847   

                                       Facility Name              City State  \
0                     121 REGIONAL DISPOSAL FACILITY           MELISSA    TX   
1                        58TH ST LF (MAIN COUNTY LF)             MIAMI    FL   
2                                     A.C.M.S., Inc.  Lake Panasoffkee    FL   
3  ACADIA PARISH POLICE JURY - ACADIA PARISH SANI...              EGAN    LA   
4                                    ADRIAN LANDFILL            ADRIAN    MI   

  Industry Type (subparts) Total reported emissions CO2 emissions  \
0                       HH                325416.25           NaN   
1                       HH                 56991.75           NaN   
2                       HH                 313745.5           NaN   
3                       HH                  46562.5           NaN   
4                       HH                  6471.75           NaN   

  Methane (CH4) emissions Nitrous Oxide (N2O) emissions SF6 emissions  \
0               325416.25                           NaN           NaN   
1                56991.75                           NaN           NaN   
2                313745.5                           NaN           NaN   
3                 46562.5                           NaN           NaN   
4                 6471.75                           NaN           NaN   

  Subpart Letter     Name of industry   Facility Type  
0             HH  Municipal Landfills  Direct Emitter  
1             HH  Municipal Landfills  Direct Emitter  
2             HH  Municipal Landfills  Direct Emitter  
3             HH  Municipal Landfills  Direct Emitter  
4             HH  Municipal Landfills  Direct Emitter

In [42]:
# now going to drop the industry type (subparts) column since it just repeats in the subpart letter column 
data = data.drop(columns = data.columns[5])

# checking to make sure 
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3905 entries, 0 to 3904
Data columns (total 13 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   Facility Id                    3905 non-null   object
 1   FRS Id                         3446 non-null   object
 2   Facility Name                  3905 non-null   object
 3   City                           3905 non-null   object
 4   State                          3905 non-null   object
 5   Total reported emissions       3905 non-null   object
 6   CO2 emissions                  3206 non-null   object
 7   Methane (CH4) emissions        3749 non-null   object
 8   Nitrous Oxide (N2O) emissions  2954 non-null   object
 9   SF6 emissions                  73 non-null     object
 10  Subpart Letter                 3905 non-null   object
 11  Name of industry               3905 non-null   object
 12  Facility Type                  3905 non-null   object
dtypes: 

In [50]:
# converting the columns that have numbers to numeric float 
data['Total reported emissions'] = pd.to_numeric(data['Total reported emissions'], errors = 'coerce')
data['CO2 emissions'] = pd.to_numeric(data['CO2 emissions'], errors = 'coerce')
data['Methane (CH4) emissions'] = pd.to_numeric(data['Methane (CH4) emissions'], errors = 'coerce')
data['Nitrous Oxide (N2O) emissions'] = pd.to_numeric(data['Nitrous Oxide (N2O) emissions'], errors = 'coerce')
data['SF6 emissions'] = pd.to_numeric(data['SF6 emissions'], errors = 'coerce')

In [51]:
# checking to make sure everything converted 
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3905 entries, 0 to 3904
Data columns (total 13 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Facility Id                    3905 non-null   object 
 1   FRS Id                         3446 non-null   object 
 2   Facility Name                  3905 non-null   object 
 3   City                           3905 non-null   object 
 4   State                          3905 non-null   object 
 5   Total reported emissions       3905 non-null   float64
 6   CO2 emissions                  3206 non-null   float64
 7   Methane (CH4) emissions        3749 non-null   float64
 8   Nitrous Oxide (N2O) emissions  2954 non-null   float64
 9   SF6 emissions                  73 non-null     float64
 10  Subpart Letter                 3905 non-null   object 
 11  Name of industry               3905 non-null   object 
 12  Facility Type                  3905 non-null   o

***Done with data cleaning!***

# **Exploratory analysis of data**

It seems like I have a total of 3,905 facility's data recorded. There are null values in the FRS Id, CO2 emissions, Methane emissions, Nitrous Oxide emissions, and SF6 emissions variables. I decided to just leave the nulls in there for now; I could possibly argue in the web app that these facilities should report further data (or something along those lines). Also, it wouldn't make sense to get rid of the nulls because some of the emissions types are for certain facilities; if I only left the rows that had non-null values in each column, it would probably only leave one facility type or leave an empty dataframe. 

Regarding the types of variables in the cleaned dataframe, the Facility Id, FRS Id, Facility Name, City, State, Subpart Letter, Name of industry, and Facility Type are noncontinuous string variables. For the Facility Id, each row is a unqiue category as the Facility Id is an Id value for each facility. The same goes for the FRS Id and Facility Name. Regarding the City and State variables, these obviously describe the various city & states that each facility is located in. There are 2,178 cities reported and 54 states (It is higher than 50 because Washington D.C., Guam, Puerto Rico, and Virgin Islands were included as 'States'.) There are 16 categories in the Subpart Letter variable, which correlate with the categories in Name of Industry. These two variables then correlate with the 6 categories in Facility Type. 


The emissions variables are continuous integers. There were 508 outliers found in the Total emisisons reported column, 431 outliers found in the CO2 emissions column, 559 outliers found in the methane emissions column, 498 outliers found in the nitrous oxide emissions column, and 3 outliers found in the SF6 emissions column. The number of outliers found in each column makes logical sense, as the standard deviation for the SF6 emission column was the lowest with also the lowest count of values in the column. Additionally, methane emissions had the highest average value, which I found interesting since it was higher than th total reported emissions average value. Though this could have been because some facilities could have had more methane emissions and less of the others, creating a 'medium' valued total reported emissions and a 'high' valued methane emissions. Since there were many observations in the dataframe (3,905), the distributions of each emissions columns were quite high; this is expected due to the large amounts of observations. 



In [54]:
# using unique function to find the categories in categorical variables 
data['Facility Id'].unique()

array([1004377, 1004680, 1012058, ..., 1012259, 1012277, 1014090],
      dtype=object)

In [63]:
array_city = data['City'].unique()
len(array_city)

2178

In [66]:
array_state = data['State'].unique()
len(array_state)

54

In [67]:
print(array_state)

['TX' 'FL' 'LA' 'MI' 'AL' 'GA' 'IL' 'MO' 'OH' 'NY' 'TN' 'CA' 'AZ' 'NC'
 'OK' 'VA' 'SC' 'MN' 'IN' 'MS' 'KY' 'MT' 'IA' 'CO' 'SD' 'NE' 'HI' 'MD'
 'WY' 'PA' 'AR' 'ND' 'NM' 'OR' 'DE' 'UT' 'NJ' 'PR' 'KS' 'MA' 'WA' 'WI'
 'GU' 'WV' 'AK' 'ID' 'NV' 'RI' 'DC' 'CT' 'NH' 'ME' 'VI' 'VT']


In [57]:
data['Subpart Letter'].unique()

array(['HH', 'FF', 'C', 'D', 'P', 'H', 'T', 'TT', 'V', 'SS', 'II', 'DD',
       'W-GB', 'W-LDC', 'W-ONSH', 'W-TRANS'], dtype=object)

In [68]:
array_letter = data['Subpart Letter'].unique()
len(array_letter)

16

In [58]:
data['Name of industry'].unique()

array(['Municipal Landfills', 'Underground Coal Mines',
       'Stationary Combustion', 'Electricity Generation',
       'Hydrogen Production', 'Cement Production', 'Magnesium Production',
       'Industrial Waste Landfills', 'Nitric Acid Production',
       'Manufacture of Electric Transmission and Distribution Equipment',
       'Industrial Wastewater Treatment', 'SF6 from Electrical Equipment',
       'Petroleum and Nautral Gas Systems - Gathering and Boosting',
       'Petroleum and Natural Gas Systems – Natural Gas Local Distribution Company',
       'Petroleum and Natural Gas Systems – Onshore Production',
       'Petroleum and Natural Gas Systems – Transmission Pipelines'],
      dtype=object)

In [69]:
array_indust = data['Name of industry'].unique()
len(array_indust)

16

In [59]:
data['Facility Type'].unique()

array(['Direct Emitter', 'SF6 from Elec. Equip.', 'Gathering & Boosting',
       'LDC - Direct Emissions', 'Onshore Oil & Gas Production',
       'Transmission Pipelines'], dtype=object)

In [70]:
array_facility = data['Facility Type'].unique()
len(array_facility)

6

In [71]:
# giving statistical descriptions for the numerical columns in df 
data.describe()

Total reported emissions  CO2 emissions  Methane (CH4) emissions  \
count              3.905000e+03   3.206000e+03             3.749000e+03   
mean               2.165676e+05   2.323134e+05             2.582948e+04   
std                7.058658e+05   7.655647e+05             7.625943e+04   
min                0.000000e+00   0.000000e+00             1.470000e-01   
25%                2.843109e+04   2.692753e+04             2.375000e+01   
50%                5.431384e+04   5.086830e+04             2.175000e+02   
75%                1.317655e+05   1.308072e+05             1.897700e+04   
max                1.586076e+07   1.573642e+07             1.187791e+06   

       Nitrous Oxide (N2O) emissions  SF6 emissions  
count                    2954.000000      73.000000  
mean                      685.847709   27929.705162  
std                      4204.960907   41411.690761  
min                         0.178800     198.565200  
25%                        18.774000    4887.271200  
50%                        38.442000   11820.796800  
75%                       118.529500   39352.822800  
max                    115954.780000  251256.000000

In [76]:
# calculating outliers for total reported emissions column 
Q1 = data['Total reported emissions'].quantile(0.25)
Q3 = data['Total reported emissions'].quantile(0.75)
IQR = Q3 - Q1

# identify outliers
threshold = 1.5
outliers = data[(data['Total reported emissions'] < Q1 - threshold * IQR) | (data['Total reported emissions'] > Q3 + threshold * IQR)]
print(len(outliers))

508


In [77]:
# calculating outliers for CO2 emissions column 
Q1 = data['CO2 emissions'].quantile(0.25)
Q3 = data['CO2 emissions'].quantile(0.75)
IQR = Q3 - Q1

# identify outliers
threshold = 1.5
outliers = data[(data['CO2 emissions'] < Q1 - threshold * IQR) | (data['CO2 emissions'] > Q3 + threshold * IQR)]
print(len(outliers))

431


In [78]:
# calculating outliers for methane emissions column 
Q1 = data['Methane (CH4) emissions'].quantile(0.25)
Q3 = data['Methane (CH4) emissions'].quantile(0.75)
IQR = Q3 - Q1

# identify outliers
threshold = 1.5
outliers = data[(data['Methane (CH4) emissions'] < Q1 - threshold * IQR) | (data['Methane (CH4) emissions'] > Q3 + threshold * IQR)]
print(len(outliers))

559


In [79]:
# calculating outliers for nitrous oxide emissions column 
Q1 = data['Nitrous Oxide (N2O) emissions'].quantile(0.25)
Q3 = data['Nitrous Oxide (N2O) emissions'].quantile(0.75)
IQR = Q3 - Q1

# identify outliers
threshold = 1.5
outliers = data[(data['Nitrous Oxide (N2O) emissions'] < Q1 - threshold * IQR) | (data['Nitrous Oxide (N2O) emissions'] > Q3 + threshold * IQR)]
print(len(outliers))

498


In [80]:
# calculating outliers for SF6 emissions column 
Q1 = data['SF6 emissions'].quantile(0.25)
Q3 = data['SF6 emissions'].quantile(0.75)
IQR = Q3 - Q1

# identify outliers
threshold = 1.5
outliers = data[(data['SF6 emissions'] < Q1 - threshold * IQR) | (data['SF6 emissions'] > Q3 + threshold * IQR)]
print(len(outliers))

3


# **Data dictionary**

* **Facility Id**: A unique identification number that has been given to each facility required to report to the Greenhouse Gas Reporting Program. 
    * Data type: Text
    * Field size: 3,905
    * Example: 1004377


* **FRS Id**: A Facility Registry System Identification Number (FRS). The system itself is a centrally managed EPA database that identifies facilities, sites or places subject to environmental regulations or of environmental interest across EPA programs.
    * Data type: Text
    * Field size: 3,905
    * Example: 10043803578


* **Facility Name**: The name given to the facility by the company who owns said facility. 
    * Data type: Text
    * Field size: 3,905
    * Example: 121 REGIONAL DISPOSAL FACILITY


* **City**: The city where the facility is located. 
    * Data type: Text
    * Field size: 2,178
    * Example: MELISSA


* **State**: The state where the facility is located in abbreviated format. 
    * Data type: Text
    * Field size: 54
    * Example: TX


* **Total reported emissions**: Emissions totalled from facilities including carbon dioxide, methane, nitrous oxide, PFCs, HFCs, SF6, NF3, other fully fluorinated GHGs, HFEs, very short-lived compounds, and other emission types (not specified). In units of metric tons CO2e using global warming potential (GWP) values provided in the IPCC's (Intergovernmental Panel on Climate Change) AR4 (Fourth Assessement Report). 
    * Data type: Integer
    * Field size: 3,905
    * Example: 325416.25


* **CO2 emissions**: Carbon dioxide emissions reported from facilities. In units of metric tons. 
    * Data type: Integer
    * Field size: 3,905
    * Example: 103134.1


* **Methane (CH4) emissions**: Methane emissions reported from facilities. In units of metric tons CO2e using global warming potential (GWP) values provided in the IPCC's AR4 report.  
    * Data type: Integer
    * Field size: 3,905
    * Example: 4818.5


* **Nitrous Oxide (N2O) emissions**: Nitrous oxide emissions reported from facilities.  In units of metric tons CO2e using global warming potential (GWP) values provided in the IPCC's AR4 report.  
    * Data type: Integer
    * Field size: 3,905
    * Example: 210.388


* **SF6 emissions**: Sulfur hexafluoride emissions reported from facilities. In units of metric tons CO2e using global warming potential (GWP) values provided in the IPCC's AR4 report. 
    * Data type: Integer
    * Field size: 3,905
    * Example: 58276.8


* **Subpart Letter**: Abbreviation used to correlate to full name of industry. 
    * Data type: Integer
    * Field size: 16 
    * 'HH' = Municipal Landfills
    * 'FF' = Underground Coal Mines
    * 'C' = Stationary Combustion
    * 'D' = Electricity Generation
    * 'P' = Hydrogen Production
    * 'H' = Cement Production
    * 'T' = Magnesium Production
    * 'TT' = Industrial Waste Landfills 
    * 'V' = Nitric Acid Production
    * 'SS' = Manufacture of Electric Transmission & Distribution Equipment 
    * 'II' = Industrial Wastewater treatment
    * 'DD' = SF6 from Electrical Equipment
    * 'W-GB' = Petroleum & Natural Gas Systems - Gathering & Boosting
    * 'W-LDC' = Petroleum & Natural Gas Systems - Natural Gas Local Distribution Company
    * 'W-ONSH' = Petroleum & Natural Gas Systems - Onshore Production
    * 'W-TRANS' = Petroleum & Natural Gas Systems - Transmission Pipelines


* **Name of industry**: Correlates with Subpart Letter column. See above for full definitions of each field. 
    * Data type: Text
    * Field size: 16




* **Facility Type**: Correlates with each industry name. 
    * Data type: Text
    * Field size: 6
    * 'Direct Emitter' = HH, FF, C, D, P, H, T, V, SS, II, 
    * 'SF6 from Elec. Equip.' = DD
    * 'Gathering & Boosting' = W-GB
    * 'LDC - Direct Emissions' = W-LDC 
    * 'Onshore Oil & Gas Production' = W-ONSH
    * 'Transmission Pipelines' = W-TRANS

# **UI and data visualization brainstorm**

***UI Ideas**: 

* Navigation bar to go through different facilities by state or facility type
* Radio buttons to click on to show different information on graph (i.e. showing total reported emissions by either the emission type or by the facility type)
* Multi-select dropdown to filter a visualization by state or name of industry
* Every transition from one graph to another or from one nav bar selection to another to be animated 

***Data visualization ideas:***

* Bar graph showing the amount of emissions (either by emissions type or facility type? - and have a radio button to select either or)
* A map of the different facilites in contiguous USA, with density dots (larger dots showing higher emitting areas & vice versa)
* Pie chart showing the amount of each emissions for one facility, state, or city (with radio buttons to select through the different options) 

***Exporting CSV***

In [81]:
data.to_csv('/Users/sophiekim/Desktop/4th year (spring)/DS 4003/project/data.csv')